lib/mayaUsd/ufe/UsdUndoMaterialCommands.cpp

```diff
@@ -375,43 +418,162 @@ void UsdUndoAssignNewMaterialCommand::redo()
         auto addMaterialCmd
             = std::dynamic_pointer_cast<MAYAUSD_NS::ufe::UsdUndoAddNewPrimCommand>(*cmdsIt++);
         auto addShaderCmd = std::dynamic_pointer_cast<UsdUndoCreateFromNodeDefCommand>(*cmdsIt);
-        connectShaderToMaterial(addShaderCmd->insertedChild(), addMaterialCmd->newPrim());
+        connectShaderToMaterial(addShaderCmd->insertedChild(), addMaterialCmd->newPrim(), _nodeId);
     }
 }
 
-void UsdUndoAssignNewMaterialCommand::connectShaderToMaterial(
-    Ufe::SceneItem::Ptr shaderItem,
-    UsdPrim             materialPrim)
+void UsdUndoAssignNewMaterialCommand::markAsFailed()
 {
-    auto                 shaderUsdItem = std::dynamic_pointer_cast<UsdSceneItem>(shaderItem);
-    auto                 shaderPrim = UsdShadeShader(shaderUsdItem->prim());
-    UsdShadeOutput       materialOutput;
-    PXR_NS::SdrRegistry& registry = PXR_NS::SdrRegistry::GetInstance();
+    _cmds->undo();
+    _cmds.reset();
+}
+
+UsdUndoAddNewMaterialCommand::UsdUndoAddNewMaterialCommand(
+    const UsdSceneItem::Ptr& parentItem,
+    const std::string&       nodeId)
+    : Ufe::InsertChildCommand()
+    , _parentPath((parentItem && parentItem->prim().IsActive()) ? parentItem->path() : Ufe::Path())
+    , _nodeId(nodeId)
+    , _cmds(std::make_shared<Ufe::CompositeUndoableCommand>())
+{
+}
+
+UsdUndoAddNewMaterialCommand::~UsdUndoAddNewMaterialCommand() { }
+
+UsdUndoAddNewMaterialCommand::Ptr
+UsdUndoAddNewMaterialCommand::create(const UsdSceneItem::Ptr& parentItem, const std::string& nodeId)
+{
+    // Changing the hierarchy of invalid items is not allowed.
+    if (!parentItem || !parentItem->prim().IsActive())
+        return nullptr;
+
+    return std::make_shared<UsdUndoAddNewMaterialCommand>(parentItem, nodeId);
+}
+
+Ufe::SceneItem::Ptr UsdUndoAddNewMaterialCommand::insertedChild() const
+{
+    if (_cmds) {
```

Comment:
There is no doc about insertChild: why it just inserts in the 2nd position of the list of commands is quite obscure to me.